accepted/0000-yarn-tasks.md

```diff
@@ -0,0 +1,103 @@
+- Start Date: 2018-11-21
+- RFC PR:
+- Yarn Issue: https://github.com/yarnpkg/yarn/issues/5773
+
+# Summary
+
+Yarn CLI command `yarn run` will be able to orchestrate various npm scripts, allowing them to run in parallel by default and matching any glob patterns.
+
+# Motivation
+
+Many build processes involve Grunt, Gulp, or other task runners that have great use cases but also require separate plugins, configuration files, and understanding of the API. `yarn run` would remove the need for installing these often global dpendencies and lower maintenance costs by providing a simpler interface defined in package.json. This would also cut down on the script run times and would be useful with workspaces as mentioned in https://github.com/yarnpkg/yarn/issues/6689.
+
+Feature Requirements:
+
+- Be able to run multiple commands in parallel across platform
+- Show and differentiate console output for each command
+- Compatible with existing behavior of `npm run` command
+
+Prior Art:
+
+- [npm-run-all](https://www.npmjs.com/package/npm-run-all)
+- [gulp](https://gulpjs.com/)
+
+# Detailed design
+
+## Simple use case
+
+// In `package.json`
+
+```js
+{
+    "scripts": {
+        "lint": "eslint .",
+        "test": "jest"
+    }
+    "tasks": {
+        "verify": ["lint", "test"]
+    }
+}
+```
+
+`yarn run verify` would run lint and test commands in parallel
+
+## Advanced use case - pattern matching
+
+```js
+{
+    "scripts": {
+        "lint": "eslint .",
+        "test:unit": "jest __tests__",
+        "test:integration": "jest __integration__"
+    }
+    "tasks": {
+        "verify": ["lint", "test**"]
+    }
+}
+```
+
+`yarn run verify` would run lint, test:unit, and test:integration commands in parallel
+
+## Advanced use case - nested tasks
+
+```js
+{
+    "scripts": {
+        "build": "webpack",
+        "lint": "eslint .",
+        "test": "jest",
+    }
+    "tasks": {
+        "verify": ["lint", "test"],
+        "dev": ["verify", "build"]
+    }
+}
+```
+
+`yarn run dev` would run lint, test, and build in parallel
+
+_NOTE_ Nested arrays or serial runs will not be supported in tasks because they can be done via scripts using `&&` operator

```

Comment:
Perhaps this is a horrible (and confusing) suggestion... what if nested arrays was how it all worked? Something like:

```js
{
  "scripts": {
    "clean": "rimraf ./build",
    "build": "webpack",
    "format": "prettier --write .",
    "lint": "eslint .",
    "test": "jest"
  },
  "tasks": {
    "verify": ["format", "lint"], // <-- Runs "format" and "lint" serially
    "pre-commit": [["verify", "test"]] // <-- Runs "verify" and "test" in parallel
    "dev": ["clean", ["build", "test --watch"]] // <-- Runs "clean", then runs "build" and "test --watch" in parallel
  }
}
```

Reply:
It is a bit confusing (lol) but it's not a horrible suggestion. I like it because it broadens the scope of `tasks` to also solve the problem of just wanting to run tasks serially without having to do `yarn lint && yarn test`. It helps keep the runner (Yarn) out of the scripts more.

The nested array syntax is a bit funky, but we do see it Babel configurations. Not sure if that helps or hurts the argument though.

Reply:
Thinking down this path further, perhaps `tasks` and `scripts` shouldn't co-exist. If you're wanting to use `tasks` then it should all be defined there. Formatting can also help show a serial path and parallel tracks:

```js
{
  "tasks": {
    "verify": [
        "prettier --write .",
        "eslint ."
    ],
    "pre-commit": [
        ["verify", "jest"]
    ],
    "dev": [
        "rimraf ./build",
        ["webpack", "jest --watch"]
    ]
  }
}
```